File: test/proto/java/Provider.java
// Test Provider (prototype) — Java port of the canonical ts/provider.ts.
//
// Reads the shared corpus (build/test/test.json) and hands test code clean,
// normalized cases. It is NOT a test runner: it never calls the subject and
// never asserts. See ../PROVIDER.md for the model and ../AGENTS.md for usage.
//
// Zero runtime dependencies (JDK standard library only) — uses the bundled
// minimal Json parser instead of Gson/Jackson.

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@SuppressWarnings({"unchecked", "rawtypes"})
public class Provider {

  public static final String NULLMARK = "__NULL__";
  public static final String UNDEFMARK = "__UNDEF__";
  public static final String EXISTSMARK = "__EXISTS__";

  // Sentinel distinguishing "key absent" from "value is null" in getpath.
  private static final Object MISSING = new Object();

  private final Object spec;

  public Provider(Object spec) {
    this.spec = spec;
  }

  // ─── loading ─────────────────────────────────────────────────────────────

  // Default corpus path resolves to build/test/test.json relative to the repo
  // root (this file lives at test/proto/java). A non-null path is used as-is,
  // so callers (e.g. the smoke harness running from the repo root) may pass an
  // absolute or repo-relative path directly.
  public static Provider load(String path) {
    String file = path != null ? path : defaultTestFile();
    try {
      String json = new String(Files.readAllBytes(Paths.get(file)), StandardCharsets.UTF_8);
      return new Provider(Json.parse(json));
    } catch (IOException e) {
      throw new RuntimeException("Failed to read corpus: " + file, e);
    }
  }

  private static String defaultTestFile() {
    // This source is at test/proto/java; the corpus is four levels up.
    Path here = Paths.get(System.getProperty("user.dir"));
    return here.resolve(Paths.get("build", "test", "test.json")).toString();
  }

  public Object raw() {
    return spec;
  }

  private Map<String, Object> root() {
    if (spec instanceof Map) {
      Object struct = ((Map<String, Object>) spec).get("struct");
      if (struct instanceof Map) {
        return (Map<String, Object>) struct;
      }
      return (Map<String, Object>) spec;
    }
    return new LinkedHashMap<>();
  }

  private Map<String, Object> fnNode(String fn) {
    Object node = null;
    if (spec instanceof Map) {
      Object struct = ((Map<String, Object>) spec).get("struct");
      if (struct instanceof Map && ((Map<String, Object>) struct).containsKey(fn)) {
        node = ((Map<String, Object>) struct).get(fn);
      } else if (((Map<String, Object>) spec).containsKey(fn)) {
        node = ((Map<String, Object>) spec).get(fn);
      }
    }
    if (node == null) {
      throw new IllegalArgumentException("Unknown function: " + fn);
    }
    return (Map<String, Object>) node;
  }

  public List<String> functions() {
    Map<String, Object> root = root();
    List<String> out = new ArrayList<>();
    for (Map.Entry<String, Object> e : root.entrySet()) {
      if (isGroupBag(e.getValue()) || hasGroups(e.getValue())) {
        out.add(e.getKey());
      }
    }
    return out;
  }

  public List<String> groups(String fn) {
    Map<String, Object> node = fnNode(fn);
    List<String> out = new ArrayList<>();
    for (Map.Entry<String, Object> e : node.entrySet()) {
      if (!"name".equals(e.getKey()) && isGroupBag(e.getValue())) {
        out.add(e.getKey());
      }
    }
    return out;
  }

  // group == null means "all groups for the function".
  public List<Entry> entries(String fn, String group) {
    Map<String, Object> node = fnNode(fn);
    List<String> groupList = group != null ? List.of(group) : groups(fn);
    List<Entry> out = new ArrayList<>();
    for (String g : groupList) {
      Object bag = node.get(g);
      if (!isGroupBag(bag)) {
        continue;
      }
      List<Object> set = (List<Object>) ((Map<String, Object>) bag).get("set");
      for (int i = 0; i < set.size(); i++) {
        out.add(normalize(fn, g, i, (Map<String, Object>) set.get(i)));
      }
    }
    return out;
  }

  // A group bag is a map with a `set` list.
  private static boolean isGroupBag(Object v) {
    return v instanceof Map && ((Map<String, Object>) v).get("set") instanceof List;
  }

  // A function node has at least one child group bag.
  private static boolean hasGroups(Object v) {
    if (!(v instanceof Map)) {
      return false;
    }
    for (Map.Entry<String, Object> e : ((Map<String, Object>) v).entrySet()) {
      if (!"name".equals(e.getKey()) && isGroupBag(e.getValue())) {
        return true;
      }
    }
    return false;
  }

  // ─── normalized records ──────────────────────────────────────────────────

  public static final class Input {
    public enum Kind {
      IN,
      ARGS,
      CTX
    }

    public Kind kind;
    public Object in;
    public List<Object> args;
    public Map<String, Object> ctx;
  }

  public static final class ErrorCheck {
    public boolean any;
    public String text;
    public boolean regex;
  }

  public static final class Expect {
    public enum Kind {
      VALUE,
      ERROR,
      MATCH,
      ABSENT
    }

    public Kind kind;
    public boolean hasValue;
    public Object value;
    public ErrorCheck error;
    public Object match;
  }

  public static final class Entry {
    public String function;
    public String group;
    public int index;
    public String id;
    public boolean doc;
    public String client;
    public Input input;
    public Expect expect;
    public Object raw;
  }

  public static final class MatchResult {
    public boolean ok;
    public List<String> path;
    public Object expected;
    public Object actual;
  }

  private static boolean has(Map<String, Object> raw, String key) {
    return raw.containsKey(key);
  }

  private static Entry normalize(String fn, String group, int index, Map<String, Object> raw) {
    Entry e = new Entry();
    e.function = fn;
    e.group = group;
    e.index = index;
    Object id = raw.get("id");
    e.id = id != null ? String.valueOf(id) : null;
    e.doc = Boolean.TRUE.equals(raw.get("doc"));
    Object client = raw.get("client");
    e.client = client != null ? String.valueOf(client) : null;
    e.input = resolveInput(raw);
    e.expect = resolveExpect(raw);
    e.raw = raw;
    return e;
  }

  private static Input resolveInput(Map<String, Object> raw) {
    Input in = new Input();
    if (has(raw, "ctx")) {
      in.kind = Input.Kind.CTX;
      in.ctx = (Map<String, Object>) raw.get("ctx");
      return in;
    }
    if (has(raw, "args")) {
      in.kind = Input.Kind.ARGS;
      in.args = (List<Object>) raw.get("args");
      return in;
    }
    in.kind = Input.Kind.IN;
    in.in = has(raw, "in") ? raw.get("in") : null;
    return in;
  }

  private static ErrorCheck parseErr(Object err) {
    ErrorCheck c = new ErrorCheck();
    if (Boolean.TRUE.equals(err)) {
      c.any = true;
      return c;
    }
    if (err instanceof String) {
      String s = (String) err;
      Matcher m = Pattern.compile("^/(.+)/$").matcher(s);
      if (m.matches()) {
        c.regex = true;
        c.text = m.group(1);
        return c;
      }
      c.text = s;
      return c;
    }
    // Non-true, non-string err spec: treat as "any error".
    c.any = true;
    return c;
  }

  private static Expect resolveExpect(Map<String, Object> raw) {
    Object matchPart = has(raw, "match") ? raw.get("match") : null;
    Expect ex = new Expect();
    if (has(raw, "err")) {
      ex.kind = Expect.Kind.ERROR;
      ex.error = parseErr(raw.get("err"));
      ex.match = matchPart;
      return ex;
    }
    // KEY PRESENCE, not null-check: "out" present even if null => VALUE.
    if (has(raw, "out")) {
      ex.kind = Expect.Kind.VALUE;
      ex.hasValue = true;
      ex.value = raw.get("out");
      ex.match = matchPart;
      return ex;
    }
    if (has(raw, "match")) {
      ex.kind = Expect.Kind.MATCH;
      ex.match = raw.get("match");
      return ex;
    }
    ex.kind = Expect.Kind.ABSENT;
    return ex;
  }

  // ─── pure comparison helpers ─────────────────────────────────────────────

  // stringify(x) = x if it is already a String, else compact JSON.
  public static String stringify(Object x) {
    return x instanceof String ? (String) x : Json.stringify(x);
  }

  private static Object normNull(Object x) {
    if (NULLMARK.equals(x) || x == null) {
      return null;
    }
    if (x instanceof List) {
      List<Object> out = new ArrayList<>();
      for (Object v : (List<Object>) x) {
        out.add(normNull(v));
      }
      return out;
    }
    if (x instanceof Map) {
      Map<String, Object> out = new LinkedHashMap<>();
      for (Map.Entry<String, Object> e : ((Map<String, Object>) x).entrySet()) {
        out.put(e.getKey(), normNull(e.getValue()));
      }
      return out;
    }
    return x;
  }

  private static Object normMark(Object x) {
    if (NULLMARK.equals(x)) {
      return null;
    }
    if (x instanceof List) {
      List<Object> out = new ArrayList<>();
      for (Object v : (List<Object>) x) {
        out.add(normMark(v));
      }
      return out;
    }
    if (x instanceof Map) {
      Map<String, Object> out = new LinkedHashMap<>();
      for (Map.Entry<String, Object> e : ((Map<String, Object>) x).entrySet()) {
        out.put(e.getKey(), normMark(e.getValue()));
      }
      return out;
    }
    return x;
  }

  public static boolean matchval(Object check, Object base) {
    if (scalarEq(check, base)) {
      return true;
    }
    if (check instanceof String) {
      String chk = (String) check;
      String basestr = stringify(base);
      Matcher rem = Pattern.compile("^/(.+)/$").matcher(chk);
      if (rem.matches()) {
        return Pattern.compile(rem.group(1)).matcher(basestr).find();
      }
      return basestr.toLowerCase().contains(chk.toLowerCase());
    }
    // A "function" check (not representable from JSON) would return true; no
    // such value arises from the parsed corpus.
    return false;
  }

  public static boolean equal(Object expected, Object actual) {
    return deepEq(normNull(expected), normNull(actual));
  }

  // Strict variant for the runner's { null: false } functions, where an absent
  // value is distinct from JSON null. Only __NULL__ is normalized.
  public static boolean equalStrict(Object expected, Object actual) {
    return deepEq(normMark(expected), normMark(actual));
  }

  // Scalar identity mirroring JS ===: distinguishes Boolean from Number, and
  // compares numbers/strings/booleans by value.
  private static boolean scalarEq(Object a, Object b) {
    if (a == b) {
      return true;
    }
    if (a == null || b == null) {
      return false;
    }
    if ((a instanceof Boolean) != (b instanceof Boolean)) {
      return false;
    }
    if (a instanceof Number && b instanceof Number) {
      return ((Number) a).doubleValue() == ((Number) b).doubleValue();
    }
    return a.equals(b);
  }

  private static boolean deepEq(Object a, Object b) {
    if (a == b) {
      return true;
    }
    if (a instanceof List && b instanceof List) {
      List<Object> la = (List<Object>) a;
      List<Object> lb = (List<Object>) b;
      if (la.size() != lb.size()) {
        return false;
      }
      for (int i = 0; i < la.size(); i++) {
        if (!deepEq(la.get(i), lb.get(i))) {
          return false;
        }
      }
      return true;
    }
    if (a instanceof List || b instanceof List) {
      return false;
    }
    if (a instanceof Map && b instanceof Map) {
      Map<String, Object> ma = (Map<String, Object>) a;
      Map<String, Object> mb = (Map<String, Object>) b;
      if (ma.size() != mb.size()) {
        return false;
      }
      for (Map.Entry<String, Object> e : ma.entrySet()) {
        if (!mb.containsKey(e.getKey()) || !deepEq(e.getValue(), mb.get(e.getKey()))) {
          return false;
        }
      }
      return true;
    }
    if (a instanceof Map || b instanceof Map) {
      return false;
    }
    return scalarEq(a, b);
  }

  public static boolean errorMatches(ErrorCheck check, String message) {
    if (check.any) {
      return true;
    }
    if (check.text == null) {
      return false;
    }
    if (check.regex) {
      return Pattern.compile(check.text).matcher(message).find();
    }
    return message.toLowerCase().contains(check.text.toLowerCase());
  }

  // Partial structural match: every leaf of `check` must match `base` at its path.
  public static MatchResult structMatch(Object check, Object base) {
    MatchResult result = new MatchResult();
    result.ok = true;
    walkLeaves(
        check,
        new ArrayList<>(),
        (val, path) -> {
          if (!result.ok) {
            return;
          }
          Object baseval = getpath(base, path);
          if (baseval != MISSING && scalarEq(val, baseval)) {
            return;
          }
          if (UNDEFMARK.equals(val) && baseval == MISSING) {
            return;
          }
          if (EXISTSMARK.equals(val) && baseval != MISSING && baseval != null) {
            return;
          }
          Object compareBase = baseval == MISSING ? null : baseval;
          if (!matchval(val, compareBase)) {
            result.ok = false;
            result.path = path;
            result.expected = val;
            result.actual = compareBase;
          }
        });
    return result;
  }

  private interface LeafFn {
    void visit(Object val, List<String> path);
  }

  private static void walkLeaves(Object node, List<String> path, LeafFn fn) {
    if (node instanceof List) {
      List<Object> l = (List<Object>) node;
      for (int i = 0; i < l.size(); i++) {
        List<String> next = new ArrayList<>(path);
        next.add(String.valueOf(i));
        walkLeaves(l.get(i), next, fn);
      }
    } else if (node instanceof Map) {
      for (Map.Entry<String, Object> e : ((Map<String, Object>) node).entrySet()) {
        List<String> next = new ArrayList<>(path);
        next.add(e.getKey());
        walkLeaves(e.getValue(), next, fn);
      }
    } else {
      fn.visit(node, path);
    }
  }

  // Returns MISSING for an absent path (distinct from a present null).
  private static Object getpath(Object store, List<String> path) {
    Object cur = store;
    for (String key : path) {
      if (cur == null || cur == MISSING) {
        return MISSING;
      }
      if (cur instanceof List) {
        List<Object> l = (List<Object>) cur;
        int idx;
        try {
          idx = Integer.parseInt(key);
        } catch (NumberFormatException nfe) {
          return MISSING;
        }
        cur = (idx >= 0 && idx < l.size()) ? l.get(idx) : MISSING;
      } else if (cur instanceof Map) {
        Map<String, Object> m = (Map<String, Object>) cur;
        cur = m.containsKey(key) ? m.get(key) : MISSING;
      } else {
        return MISSING;
      }
    }
    return cur;
  }
}

File: test/proto/haskell/Provider.hs
-- Test Provider (prototype) — Haskell port of the canonical ts/provider.ts.
--
-- Reads the shared corpus (build/test/test.json) and hands test code clean,
-- normalized cases. It is NOT a test runner: it never calls the subject and
-- never asserts. See ../PROVIDER.md for the model and ../AGENTS.md for usage.
--
-- DEPENDENCY-FREE: GHC base only. The JSON reader is self-contained (adapted
-- from haskell/test/Runner.hs), producing an order-preserving `Json` type
-- whose objects are association lists.
--
-- NOTE: GHC is unavailable in this environment, so this file has NOT been
-- compiled or executed. It is a faithful port written from the canonical
-- TypeScript source and the PROVIDER.md spec.

{-# LANGUAGE LambdaCase #-}

module Provider
  ( Json(..)
  , Provider
  , Entry(..)
  , Input(..)
  , InputKind(..)
  , Expect(..)
  , ExpectKind(..)
  , ErrorCheck(..)
  , MatchResult(..)
  , parseJson
  , load
  , raw
  , functions
  , groups
  , entries
  , stringify
  , matchval
  , equal
  , equalStrict
  , structMatch
  , errorMatches
  ) where

import Data.Char (chr, toLower)
import Data.IORef
import Data.List (isInfixOf)
import Numeric (readHex, showHex)
import System.IO.Unsafe (unsafePerformIO)

-- ─── JSON model ─────────────────────────────────────────────────────────────
-- JObj is an association list to preserve key order (mirrors JSON.parse order
-- preservation and the runner's order-preserving maps).

data Json
  = JNull
  | JBool Bool
  | JNum Double
  | JStr String
  | JArr [Json]
  | JObj [(String, Json)]
  deriving (Show)

-- Reference equality for the model: structural, order-insensitive for objects
-- (object equality compares by key membership, matching deepEq in the canon).
instance Eq Json where
  JNull        == JNull        = True
  (JBool a)    == (JBool b)    = a == b
  (JNum a)     == (JNum b)     = a == b
  (JStr a)     == (JStr b)     = a == b
  (JArr a)     == (JArr b)     = length a == length b && and (zipWith (==) a b)
  (JObj a)     == (JObj b)     =
    length a == length b
      && all (\(k, v) -> case lookup k b of Just w -> v == w; Nothing -> False) a
  _            == _            = False

-- ─── Sentinels ──────────────────────────────────────────────────────────────

nullmark, undefmark, existsmark :: String
nullmark = "__NULL__"
undefmark = "__UNDEF__"
existsmark = "__EXISTS__"

-- ─── JSON parser ────────────────────────────────────────────────────────────
-- Adapted from haskell/test/Runner.hs's IORef-backed reader, retargeted at the
-- pure order-preserving `Json` type. Handles escapes/\uXXXX, numbers,
-- true/false/null. Uses an IORef cursor over the input held in unsafePerformIO
-- so the public surface stays pure.

parseJson :: String -> Json
parseJson s0 = unsafePerformIO (jsonRead s0)
{-# NOINLINE parseJson #-}

jsonRead :: String -> IO Json
jsonRead s0 = do
  posRef <- newIORef 0
  let arr = s0
      n = length arr
      at i = arr !! i
      peek = do p <- readIORef posRef; return (if p < n then Just (at p) else Nothing)
      adv = modifyIORef' posRef (+ 1)
      skipWs = do
        p <- readIORef posRef
        if p < n && (at p `elem` " \t\n\r") then adv >> skipWs else return ()
      pval = do
        skipWs
        mc <- peek
        case mc of
          Just '{' -> pobj
          Just '[' -> parr
          Just '"' -> JStr <$> pstr
          Just 't' -> modifyIORef' posRef (+ 4) >> return (JBool True)
          Just 'f' -> modifyIORef' posRef (+ 5) >> return (JBool False)
          Just 'n' -> modifyIORef' posRef (+ 4) >> return JNull
          _        -> pnum
      pobj = do
        adv; skipWs
        mc <- peek
        if mc == Just '}' then adv >> return (JObj [])
        else do
          accRef <- newIORef []
          let loop = do
                skipWs
                k <- pstr
                skipWs; adv  -- ':'
                v <- pval
                modifyIORef' accRef ((k, v) :)
                skipWs
                c <- peek >>= \case Just ch -> adv >> return ch; Nothing -> return '}'
                if c == ',' then loop
                else do acc <- readIORef accRef; return (JObj (reverse acc))
          loop
      parr = do
        adv; skipWs
        mc <- peek
        if mc == Just ']' then adv >> return (JArr [])
        else do
          accRef <- newIORef []
          let loop = do
                v <- pval
                modifyIORef' accRef (v :)
                skipWs
                c <- peek >>= \case Just ch -> adv >> return ch; Nothing -> return ']'
                if c == ',' then loop
                else do acc <- readIORef accRef; return (JArr (reverse acc))
          loop
      pstr = do
        adv  -- opening quote
        bRef <- newIORef []
        let push c = modifyIORef' bRef (c :)
            loop = do
              p <- readIORef posRef
              let c = at p
              adv
              if c == '"' then do b <- readIORef bRef; return (reverse b)
              else if c == '\\' then do
                p2 <- readIORef posRef
                let e = at p2
                adv
                case e of
                  '"'  -> push '"' >> loop
                  '\\' -> push '\\' >> loop
                  '/'  -> push '/' >> loop
                  'n'  -> push '\n' >> loop
                  't'  -> push '\t' >> loop
                  'r'  -> push '\r' >> loop
                  'b'  -> push '\b' >> loop
                  'f'  -> push '\f' >> loop
                  'u'  -> do
                    pp <- readIORef posRef
                    let hex = take 4 (drop pp arr)
                    modifyIORef' posRef (+ 4)
                    case readHex hex of
                      [(code, _)] -> push (chr code) >> loop
                      _           -> loop
                  _    -> push e >> loop
              else push c >> loop
        loop
      pnum = do
        start <- readIORef posRef
        let go = do
              p <- readIORef posRef
              if p < n && (at p `elem` "0123456789-+.eE") then adv >> go else return ()
        go
        end <- readIORef posRef
        let tok = take (end - start) (drop start arr)
        return (JNum (read tok))
  pval

-- ─── Provider ───────────────────────────────────────────────────────────────

newtype Provider = Provider { spec :: Json }

-- Default corpus path: build/test/test.json relative to the repo root.
-- This file lives at test/proto/haskell, so the repo root is three levels up.
defaultTestFile :: FilePath
defaultTestFile = "../../../build/test/test.json"

load :: Maybe FilePath -> IO Provider
load mfile = do
  let file = maybe defaultTestFile id mfile
  contents <- readFile file
  parsed <- jsonRead contents
  return (Provider parsed)

raw :: Provider -> Json
raw = spec

-- The root used for function discovery: spec.struct if present, else spec.
specRoot :: Provider -> Json
specRoot p = case objLookup (spec p) "struct" of
  Just node -> node
  Nothing   -> spec p

fnNode :: Provider -> String -> Json
fnNode p fn =
  case objLookup (spec p) "struct" >>= \root -> objLookup root fn of
    Just node -> node
    Nothing   -> case objLookup (spec p) fn of
      Just node -> node
      Nothing   -> error ("Unknown function: " ++ fn)

functions :: Provider -> [String]
functions p =
  case specRoot p of
    JObj kvs -> [k | (k, v) <- kvs, isGroupBag v || hasGroups v]
    _        -> []

groups :: Provider -> String -> [String]
groups p fn =
  case fnNode p fn of
    JObj kvs -> [k | (k, v) <- kvs, k /= "name", isGroupBag v]
    _        -> []

entries :: Provider -> String -> Maybe String -> [Entry]
entries p fn mgroup =
  let node = fnNode p fn
      gs = maybe (groups p fn) (: []) mgroup
      collect g =
        case objLookup node g of
          Just bag | isGroupBag bag ->
            case objLookup bag "set" of
              Just (JArr items) -> zipWith (normalize fn g) [0 ..] items
              _                 -> []
          _ -> []
  in concatMap collect gs

-- A group bag is an object with a `set` array.
isGroupBag :: Json -> Bool
isGroupBag v = case v of
  JObj kvs -> case lookup "set" kvs of Just (JArr _) -> True; _ -> False
  _        -> False

-- A function node has at least one child group bag.
hasGroups :: Json -> Bool
hasGroups v = case v of
  JObj kvs -> any (\(k, x) -> k /= "name" && isGroupBag x) kvs
  _        -> False

-- ─── Entry model ────────────────────────────────────────────────────────────

data InputKind = KIn | KArgs | KCtx deriving (Eq, Show)

data Input = Input
  { ikind  :: InputKind
  , ivalue :: Json
  } deriving (Show)

data ExpectKind = EValue | EError | EMatch | EAbsent deriving (Eq, Show)

data ErrorCheck = ErrorCheck
  { anyErr :: Bool
  , etext  :: Maybe String
  , eregex :: Bool
  } deriving (Show)

data Expect = Expect
  { ekind  :: ExpectKind
  , evalue :: Maybe Json
  , eerror :: Maybe ErrorCheck
  , ematch :: Maybe Json
  } deriving (Show)

data Entry = Entry
  { function :: String
  , group    :: String
  , index    :: Int
  , eid      :: Maybe String
  , doc      :: Bool
  , client   :: Maybe String
  , input    :: Input
  , expect   :: Expect
  , eraw     :: Json
  } deriving (Show)

-- ─── Normalization (exactly per PROVIDER.md) ───────────────────────────────

-- Object key lookup preserving order semantics; Nothing when key absent.
objLookup :: Json -> String -> Maybe Json
objLookup (JObj kvs) k = lookup k kvs
objLookup _ _          = Nothing

-- Key presence on an object (the assoc-list keys), NOT a null-check.
objHas :: Json -> String -> Bool
objHas (JObj kvs) k = any ((== k) . fst) kvs
objHas _ _          = False

normalize :: String -> String -> Int -> Json -> Entry
normalize fn g idx rawEntry = Entry
  { function = fn
  , group    = g
  , index    = idx
  , eid      = case objLookup rawEntry "id" of
                 Just v | not (isNullJson v) -> Just (jsonToString v)
                 _                           -> Nothing
  , doc      = case objLookup rawEntry "doc" of Just (JBool True) -> True; _ -> False
  , client   = case objLookup rawEntry "client" of
                 Just v | not (isNullJson v) -> Just (jsonToString v)
                 _                           -> Nothing
  , input    = resolveInput rawEntry
  , expect   = resolveExpect rawEntry
  , eraw     = rawEntry
  }
  where
    isNullJson JNull = True
    isNullJson _     = False
    -- id/client are coerced to string (mirrors String(raw.id)); for the corpus
    -- these are already strings.
    jsonToString (JStr s) = s
    jsonToString v        = stringify v

resolveInput :: Json -> Input
resolveInput rawEntry
  | objHas rawEntry "ctx"  = Input KCtx  (must "ctx")
  | objHas rawEntry "args" = Input KArgs (must "args")
  | otherwise              = Input KIn  (if objHas rawEntry "in" then must "in" else JNull)
  where
    must k = maybe JNull id (objLookup rawEntry k)

parseErr :: Json -> ErrorCheck
parseErr err = case err of
  JBool True -> ErrorCheck True Nothing False
  JStr s     -> case regexInner s of
                  Just inner -> ErrorCheck False (Just inner) True
                  Nothing    -> ErrorCheck False (Just s) False
  -- Non-true, non-string err spec: treat as "any error".
  _          -> ErrorCheck True Nothing False

-- Match "/…/" — leading and trailing slash with at least one inner char.
regexInner :: String -> Maybe String
regexInner s
  | length s >= 3 && head s == '/' && last s == '/' =
      Just (take (length s - 2) (drop 1 s))
  | otherwise = Nothing

resolveExpect :: Json -> Expect
resolveExpect rawEntry
  | objHas rawEntry "err" =
      Expect EError Nothing (Just (parseErr (forceLookup "err"))) matchPart
  | objHas rawEntry "out" =
      Expect EValue (Just (forceLookup "out")) Nothing matchPart
  | objHas rawEntry "match" =
      Expect EMatch Nothing Nothing (Just (forceLookup "match"))
  | otherwise =
      Expect EAbsent Nothing Nothing Nothing
  where
    matchPart = if objHas rawEntry "match" then objLookup rawEntry "match" else Nothing
    forceLookup k = maybe JNull id (objLookup rawEntry k)

-- ─── Pure comparison helpers (PROVIDER.md §5) ──────────────────────────────

-- stringify(x) = the string if it is already a string, else compact JSON.
stringify :: Json -> String
stringify (JStr s) = s
stringify v        = compactJson v

compactJson :: Json -> String
compactJson v = case v of
  JNull    -> "null"
  JBool b  -> if b then "true" else "false"
  JNum d   -> showNum d
  JStr s   -> showJsonString s
  JArr xs  -> "[" ++ intercalate "," (map compactJson xs) ++ "]"
  JObj kvs -> "{" ++ intercalate "," (map kv kvs) ++ "}"
  where
    kv (k, x) = showJsonString k ++ ":" ++ compactJson x

-- Compact number rendering close to JSON: integral doubles print without a
-- trailing ".0" (mirrors JSON.stringify).
showNum :: Double -> String
showNum d
  | isNaN d || isInfinite d = "null"
  | d == fromIntegral i     = show i
  | otherwise               = show d
  where i = round d :: Integer

showJsonString :: String -> String
showJsonString s = "\"" ++ concatMap esc s ++ "\""
  where
    esc c = case c of
      '"'  -> "\\\""
      '\\' -> "\\\\"
      '\n' -> "\\n"
      '\t' -> "\\t"
      '\r' -> "\\r"
      '\b' -> "\\b"
      '\f' -> "\\f"
      _ | c < ' ' -> "\\u" ++ pad4 (showHex (fromEnum c) "")
        | otherwise -> [c]
    pad4 h = replicate (4 - length h) '0' ++ h

intercalate :: String -> [String] -> String
intercalate _ []     = ""
intercalate _ [x]    = x
intercalate sep (x:xs) = x ++ sep ++ intercalate sep xs

-- normNull: __NULL__ → null, recurse through arrays/objects (mirrors normNull).
normNull :: Json -> Json
normNull x = case x of
  JStr s | s == nullmark -> JNull
  JArr xs                -> JArr (map normNull xs)
  JObj kvs               -> JObj [(k, normNull v) | (k, v) <- kvs]
  _                      -> x

-- normMark: only __NULL__ → null (strict variant).
normMark :: Json -> Json
normMark x = case x of
  JStr s | s == nullmark -> JNull
  JArr xs                -> JArr (map normMark xs)
  JObj kvs               -> JObj [(k, normMark v) | (k, v) <- kvs]
  _                      -> x

-- matchval(check, base): check == base; else if check is a string:
--   "/re/" ⇒ regex test over stringify(base) (PROTOTYPE: regex simplified),
--   otherwise stringify(base) contains check, case-insensitively.
-- (No function case in the Json model — JFunc does not exist here.)
matchval :: Json -> Json -> Bool
matchval check base
  | check == base = True
  | otherwise = case check of
      JStr cs ->
        let basestr = stringify base
        in case regexInner cs of
             Just re -> regexLike re basestr
             Nothing -> map toLower cs `isInfixOf` map toLower basestr
      _ -> False

-- PROTOTYPE: regex simplified. `base` has no regex engine; we approximate a
-- "/re/" check by case-sensitive substring containment of the inner pattern.
-- This is intentionally weaker than a real RegExp and is documented as such.
regexLike :: String -> String -> Bool
regexLike re hay = re `isInfixOf` hay

equal :: Json -> Json -> Bool
equal expected actual = deepEq (normNull expected) (normNull actual)

equalStrict :: Json -> Json -> Bool
equalStrict expected actual = deepEq (normMark expected) (normMark actual)

-- deepEq mirrors the canonical deepEq: arrays by length+elementwise; objects by
-- key count + membership; primitives by value. (Eq Json already implements
-- these semantics, including bool/number distinction via constructors.)
deepEq :: Json -> Json -> Bool
deepEq = (==)

errorMatches :: ErrorCheck -> String -> Bool
errorMatches check message
  | anyErr check = True
  | otherwise = case etext check of
      Nothing   -> False
      Just text ->
        if eregex check
          then regexLike text message               -- PROTOTYPE: regex simplified
          else map toLower text `isInfixOf` map toLower message

-- ─── structMatch ────────────────────────────────────────────────────────────

data MatchResult = MatchResult
  { ok        :: Bool
  , mpath     :: [String]
  , mexpected :: Maybe Json
  , mactual   :: Maybe Json
  } deriving (Show)

okResult :: MatchResult
okResult = MatchResult True [] Nothing Nothing

-- Partial structural match: every leaf of `check` must match `base` at its path.
-- Mirrors the canonical structMatch, with `__UNDEF__` requiring absence and
-- `__EXISTS__` requiring presence. First failure returns its path + values.
structMatch :: Json -> Json -> MatchResult
structMatch check base = go (walkLeaves check [])
  where
    go [] = okResult
    go ((val, path) : rest) =
      let baseval = getpath base path  -- Maybe Json; Nothing == undefined/absent
      in if leafOk val baseval
           then go rest
           else MatchResult False path (Just val) baseval
    leafOk val baseval =
      case baseval of
        Just bv | val == bv -> True
        _ ->
          case val of
            JStr s | s == undefmark -> baseval == Nothing
            JStr s | s == existsmark ->
              case baseval of Just bv -> not (isNullJson bv); Nothing -> False
            _ -> matchval val (maybe JNull id baseval)
    isNullJson JNull = True
    isNullJson _     = False

-- Walk to leaves, collecting (leafValue, path). Objects and non-empty arrays
-- recurse; scalars (and—per the canonical isNode—nulls/bools/numbers/strings)
-- are leaves. Empty containers are NOT leaves (matching walkLeaves: an empty
-- object/array yields no leaves), consistent with the canon.
walkLeaves :: Json -> [String] -> [(Json, [String])]
walkLeaves node path = case node of
  JArr xs  -> concat (zipWith (\i v -> walkLeaves v (path ++ [show i])) [0 :: Int ..] xs)
  JObj kvs -> concatMap (\(k, v) -> walkLeaves v (path ++ [k])) kvs
  _        -> [(node, path)]

-- getpath over the Json model: Nothing means undefined (absent or off the end).
getpath :: Json -> [String] -> Maybe Json
getpath store [] = Just store
getpath store (key : rest) = case store of
  JNull   -> Nothing
  JArr xs -> case readIndex key of
               Just i | i >= 0 && i < length xs -> getpath (xs !! i) rest
               _                                -> Nothing
  JObj kvs -> case lookup key kvs of
                Just v  -> getpath v rest
                Nothing -> Nothing
  _ -> Nothing

readIndex :: String -> Maybe Int
readIndex s = case reads s of [(i, "")] -> Just i; _ -> Nothing
